client/comms/wsconn.go

```diff
@@ -0,0 +1,377 @@
+func (conn *WsConn) setConnected(connected bool) {
+	conn.connectedMtx.Lock()
+	conn.connected = connected
+	conn.connectedMtx.Unlock()
+}
+
+// logRoute logs a request keyed by its id.
+func (conn *WsConn) logRequest(id uint64, req *msgjson.Message) {
+	conn.reqMtx.Lock()
+	conn.req[id] = req
+	conn.reqMtx.Unlock()
+}
+
+// FetchRequest fetches the request associated with the id. The returned
+// request is removed from the cache.
+func (conn *WsConn) FetchRequest(id uint64) (*msgjson.Message, error) {
+	conn.reqMtx.Lock()
+	defer conn.reqMtx.Unlock()
+	req := conn.req[id]
+	if req == nil {
+		return nil, fmt.Errorf("no request found for id %d", id)
+	}
+
+	delete(conn.req, id)
+
+	return req, nil
+}
+
+// NextID returns the next request id.
+func (conn *WsConn) NextID() uint64 {
+	return atomic.AddUint64(&conn.rID, 1)
+}
+
+// connect attempts to establish a websocket connection.
+func (conn *WsConn) connect() error {
+	scheme := "ws"
+
+	if conn.tlsCfg != nil {
+		scheme = "wss"
+	}
+
+	url := url.URL{
+		Scheme: scheme,
+		Host:   conn.cfg.Host,
+		Path:   conn.cfg.Path,
+	}
+
+	dialer := &websocket.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: 10 * time.Second,
+	}
+
+	if conn.tlsCfg != nil {
+		dialer.TLSClientConfig = conn.tlsCfg
+	}
+
+	var err error
+	conn.ws, _, err = dialer.Dial(url.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	conn.setConnected(true)
+	conn.ws.SetPingHandler(func(string) error {
+		now := time.Now()
+		err := conn.ws.SetReadDeadline(now.Add(conn.cfg.PingWait))
+		if err != nil {
+			log.Errorf("unable to set read deadline: %v", err)
+			return err
+		}
+
+		// Respond with a pong.
+		return conn.ws.WriteControl(websocket.PongMessage, []byte{},
+			now.Add(conn.cfg.WriteWait))
+	})
+
+	return nil
+}
+
+// keepAlive maintains an active websocket connection by reconnecting when
+// the established connection is broken. This should be run as a goroutine.
+func (conn *WsConn) keepAlive() {
+	ticker := time.NewTicker(time.Second)
+
+	for {
+		select {
+		case <-ticker.C:
+			if !conn.isConnected() {
+				err := conn.connect()
+				if err != nil {
+					log.Errorf("unable to establish a connection: %v", err)
+				}
+
+				// Run the reconnect sync after reestablishing a connection.
+				if conn.cfg.ReconnectSync != nil {
+					conn.cfg.ReconnectSync()
+				}
+			}
+
+		case <-conn.cfg.Ctx.Done():
+			now := time.Now()
+			closeMsg := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")
+			conn.writeMtx.Lock()
+			err := conn.ws.WriteControl(websocket.CloseMessage, closeMsg,
+				now.Add(conn.cfg.WriteWait))
+			conn.writeMtx.Unlock()
+			if err != nil {
+				log.Errorf("unable to close websocket connection: %v", err)
+			}
+			ticker.Stop()
+			conn.wg.Done()
+			return
+		}
+	}
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (conn *WsConn) read() {
+	msg := new(msgjson.Message)
+
```

Comment:
Please do a `defer conn.wg.Done()` here, and use a `conn.wg.Add(3)` in `Run`. Or was there a reason why that doesn't work?

Reply:
Yes, `ReadJSON` blocks until there is a message recieved, because of this the func cannot process signals effectively. Any signal recieved while `ReadJSON` is blocking will delay in processing until it unblocks which in turn would hold the wait group and disrupt shutdowns.

The way around this is to use a close message to terminate the goroutine. This causes the read to error which immediately unblocks. Non terminating read error cases (timeouts for example) sleep for a moment to avoid busy waiting until the connection is reestablished.

Reply:
The read goroutine can't be allowed to go zombie on shutdown.  This is why I was previously pushing for check on ctx.Err() or Done() so that any read error (even the continue cases) can lead to a shutdown if it is time.  To ensure it is timely, one of the other two goroutines will need to kill the connection so read ReadJSON won't continue to block.

Reply:
Note that you can check the ctx state before handling the err from ReadJSON, as one way of avoiding the sleep.  This is something that needs a solution though.

Reply:
However, proper synchronization with `connect()` can be done to avoid the sleeps completely, which is desirable. e.g. a waitForConnect() method could be created that receives on a channel that is closed when connect completes.  This channel would of course need to be protected by a mutex, but that's one way.

Reply:
working through the issues here, will get in touch for more detail if needed.